controllers/providers/pg/applier.go

```diff
@@ -1043,346 +1039,46 @@ func reconcilePGBackRestRepos(
+	if err != nil {
+		return []crunchyv1beta1.PGBackRestRepo{}, map[string]string{}, []byte{},
+			errors.Join(err, errors.New("failed to reconcile repos"))
 	}
 
-	// Add on-demand backups whose backup storage doesn't have a schedule
-	// defined
-	// XXX some of these backups might be fake, see the XXX comment in the
-	// reconcilePGBackupsSpec function for more context. Be very careful if you
-	// decide to use fields other that the Spec.BackupStorageName.
-	for _, backupRequest := range backupRequests {
-		// Backup storage already in repos, skip
-		if _, ok := backupStoragesInRepos[backupRequest.Spec.BackupStorageName]; ok {
-			continue
-		}
-
-		backupStorage, ok := backupStorages[backupRequest.Spec.BackupStorageName]
-		if !ok {
-			return []crunchyv1beta1.PGBackRestRepo{}, map[string]string{}, []byte{}, fmt.Errorf("unknown backup storage %s", backupRequest.Spec.BackupStorageName)
-		}
-
-		if len(availableRepoNames) == 0 {
-			return []crunchyv1beta1.PGBackRestRepo{}, map[string]string{}, []byte{}, errors.New("exceeded max number of repos")
-		}
-
-		repoName := availableRepoNames[0]
-		availableRepoNames = removeRepoName(availableRepoNames, repoName)
-		repo, err := genPGBackrestRepo(repoName, backupStorage, nil)
-		if err != nil {
-			return []crunchyv1beta1.PGBackRestRepo{}, map[string]string{}, []byte{}, err
-		}
-		reposReconciled.PushBack(repo)
-
-		// Keep track of backup storages which are already in use by a repo
-		backupStoragesInRepos[backupRequest.Spec.BackupStorageName] = struct{}{}
-
-		pgBackRestGlobal[fmt.Sprintf(pgBackRestPathTmpl, repo.Name)] = "/" + common.BackupStoragePrefix(db)
-		if verify := pointer.Get(backupStorage.VerifyTLS); !verify {
-			// See: https://pgbackrest.org/configuration.html#section-repository/option-repo-storage-verify-tls
-			pgBackRestGlobal[fmt.Sprintf(pgBackRestStorageVerifyTmpl, repo.Name)] = "n"
-		}
-		if forcePathStyle := pointer.Get(backupStorages[repo.Name].ForcePathStyle); !forcePathStyle {
-			// See: https://pgbackrest.org/configuration.html#section-repository/option-repo-s3-uri-style
-			pgBackRestGlobal[fmt.Sprintf(pgBackRestStorageForcePathTmpl, repo.Name)] = pgBackRestStoragePathStyle
-		}
-		sType, err := backupStorageTypeFromBackrestRepo(repo)
-		if err != nil {
-			return []crunchyv1beta1.PGBackRestRepo{},
-				map[string]string{},
-				[]byte{},
-				err
-		}
-
-		err = addBackupStorageCredentialsToPGBackrestSecretIni(sType, pgBackRestSecretIni, repo.Name, backupStoragesSecrets[backupRequest.Spec.BackupStorageName])
-		if err != nil {
-			return []crunchyv1beta1.PGBackRestRepo{},
-				map[string]string{},
-				[]byte{},
-				errors.Join(err, errors.New("failed to add backup storage credentials to PGBackrest secret data"))
-		}
+	err = reposReconciler.reconcileBackups(backups, backupStorages, backupStoragesSecrets, db)
+	if err != nil {
+		return []crunchyv1beta1.PGBackRestRepo{}, map[string]string{}, []byte{}, err
```

Comment:
besides the refactoring, the only change is that now we call `reconcileBackups` before `addNewSchedules` but not after it.